warehouse/bigquery.go

```diff
@@ -0,0 +1,394 @@
+
+	loader := bq.bqClient.Dataset(bq.conf.BigQuery.Dataset).Table(partitionTable).LoaderFrom(gcsRef)
+	loader.CreateDisposition = bigquery.CreateNever
+	if start.Equal(start.Truncate(24 * time.Hour)) {
+		// this is the first file of the partition, truncate the partition in case there is leftover data from previous failed loads
+		log.Printf("Detected first bundle of the day (start: %s), using WriteTruncate to replace any existing data in partition", start)
+		loader.WriteDisposition = bigquery.WriteTruncate
+	}
+
+	// start and wait on loading job
+	job, err := loader.Run(bq.ctx)
+	if err != nil {
+		log.Printf("Could not start BQ load job for file %s", filename)
+		return err
+	}
+	status, err := job.Wait(bq.ctx)
+	if err != nil {
+		log.Printf("Waiting on BQ load job for file %s failed", filename)
+		return err
+	}
+	if status.Err() != nil {
+		log.Printf("BQ load job for file %s failed", filename)
+		logJobErrors(status)
+		return status.Err()
+	}
+
+	return nil
+}
+
+func (bq *BigQuery) ValueToString(val interface{}, f Field) string {
+	s := fmt.Sprintf("%v", val)
+	if f.IsTime {
+		t, _ := time.Parse(time.RFC3339Nano, s)
+		return t.Format(time.RFC3339)
+	}
+
+	s = strings.Replace(s, "\n", " ", -1)
+	s = strings.Replace(s, "\x00", "", -1)
+
+	if len(s) >= bq.conf.BigQuery.VarCharMax {
+		s = s[:bq.conf.BigQuery.VarCharMax-1]
+	}
+	return s
+}
+
+func (bq *BigQuery) connectToBQ() error {
+	var err error
+	bq.ctx = context.Background()
+	bq.bqClient, err = bigquery.NewClient(bq.ctx, bq.conf.BigQuery.Project)
+	if err != nil {
+		log.Printf("Could not connect to BigQuery: %s", err)
+		return err
+	}
+	return nil
+}
+
+func (bq *BigQuery) doesTableExist(name string) bool {
+	log.Printf("checking if table %s exists", name)
+
+	q := fmt.Sprintf("SELECT COUNT(*) FROM %s.__TABLES_SUMMARY__ WHERE table_id = '%s';", bq.conf.BigQuery.Dataset, name)
+	query := bq.bqClient.Query(q)
+	query.QueryConfig.UseStandardSQL = true
+
+	iter, err := query.Read(bq.ctx)
+	if err != nil {
+		// something is horribly wrong...just give up
+		log.Fatal(err)
+	}
+
+	var row []bigquery.Value
+	err = iter.Next(&row)
+	if err != nil {
+		// not checking for iterator.Done, as count queries should always return at least 1 row
+		log.Fatal(err)
+	}
+
+	cnt := row[0].(int64)
+	return cnt > 0
+}
+
+func (bq *BigQuery) createSyncTable() error {
+	log.Printf("Creating table %s", bq.conf.BigQuery.SyncTable)
+
+	schema, err := bigquery.InferSchema(syncTable{})
+	if err != nil {
+		return err
+	}
+
+	table := bq.bqClient.Dataset(bq.conf.BigQuery.Dataset).Table(bq.conf.BigQuery.SyncTable)
+	if err := table.Create(bq.ctx, schema); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (bq *BigQuery) createExportTable() error {
+	log.Printf("Creating table %s", bq.conf.BigQuery.ExportTable)
+
+	schema, err := bigquery.InferSchema(exportSchema{})
+	if err != nil {
+		return err
+	}
+
+	// only EventStart and EventType should be required
+	for i := 2; i < len(schema); i++ {
```

Comment:
This is kinda dirty since it's order dependent...it would be cleaner to iterate through the schema object, and if the field doesn't match EventStart or EventType, then mark if not-required.